source/v5/mqtt3_to_mqtt5_adapter.c

```diff
@@ -1651,7 +1681,549 @@ static uint16_t s_aws_mqtt_client_connection_5_publish(
+        adapter_to_release = subscribe_op->base.adapter;
+    }
+
+    /* We're going away before our MQTT5 operation, make sure it doesn't try to call us back when it completes */
+    if (subscribe_op->subscribe_op != NULL) {
+        subscribe_op->subscribe_op->completion_options.completion_callback = NULL;
+        subscribe_op->subscribe_op->completion_options.completion_user_data = NULL;
+    }
+
+    aws_mqtt5_operation_release(&subscribe_op->subscribe_op->base);
+
+    aws_mem_release(operation->allocator, operation);
+
+    if (adapter_to_release != NULL) {
+        aws_ref_count_release(&adapter_to_release->internal_refs);
+    }
+}
+
+static enum aws_mqtt_qos s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(
+    enum aws_mqtt5_suback_reason_code reason_code) {
+    switch (reason_code) {
+        case AWS_MQTT5_SARC_GRANTED_QOS_0:
+        case AWS_MQTT5_SARC_GRANTED_QOS_1:
+        case AWS_MQTT5_SARC_GRANTED_QOS_2:
+            return (enum aws_mqtt_qos)reason_code;
+
+        default:
+            return AWS_MQTT_QOS_FAILURE;
+    }
+}
+
+static void s_aws_mqtt3_to_mqtt5_adapter_subscribe_completion_fn(
+    const struct aws_mqtt5_packet_suback_view *suback,
+    int error_code,
+    void *complete_ctx) {
+    (void)suback;
+
+    struct aws_mqtt3_to_mqtt5_adapter_operation_subscribe *subscribe_op = complete_ctx;
+    struct aws_mqtt_client_connection_5_impl *adapter = subscribe_op->base.adapter;
+
+    if (subscribe_op->on_suback != NULL) {
+        struct aws_byte_cursor topic_filter;
+        AWS_ZERO_STRUCT(topic_filter);
+        enum aws_mqtt_qos granted_qos = AWS_MQTT_QOS_AT_MOST_ONCE;
+
+        size_t subscription_count = aws_array_list_length(&subscribe_op->subscriptions);
+        if (subscription_count > 0) {
+            struct aws_mqtt_subscription_set_subscription_record *record = NULL;
+            aws_array_list_get_at(&subscribe_op->subscriptions, &record, 0);
+
+            topic_filter = record->subscription_view.topic_filter;
+        }
+
+        if (suback->reason_code_count > 0) {
+            granted_qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[0]);
+        }
+
+        (*subscribe_op->on_suback)(
+            &adapter->base,
+            subscribe_op->base.id,
+            &topic_filter,
+            granted_qos,
+            error_code,
+            subscribe_op->on_suback_user_data);
+    }
+
+    if (subscribe_op->on_multi_suback != NULL) {
+        AWS_VARIABLE_LENGTH_ARRAY(
+            struct aws_mqtt_topic_subscription, multi_sub_subscription_buf, suback->reason_code_count);
+        AWS_VARIABLE_LENGTH_ARRAY(
+            struct aws_mqtt_topic_subscription *, multi_sub_subscription_ptr_buf, suback->reason_code_count);
+        struct aws_mqtt_topic_subscription *subscription_ptr =
+            (struct aws_mqtt_topic_subscription *)multi_sub_subscription_buf;
+
+        struct aws_array_list multi_sub_list;
+        aws_array_list_init_static(
```

Comment:
Yikes, this was a rabbit hole. Will have to take note to use this pattern in the future if I need a temp array list >,<